### ISSUE TITLE
OSD daemon logs are not being collected

### ISSUE THREAD

**Issue by travisn** [rook/rook#2479]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Since the decoupling of the ceph version, the ceph-osd log output is not being captured in the k8s logs. In the osd logs we see the output of ceph-volume and that the osd process is started, but nothing after that. It seems to be an issue with rook starting the ceph-osd process and not capturing the stderr output. 

**Expected behavior:**
OSD logging should be captured.

**How to reproduce it (minimal and precise):**

- Create the rook operator
- Create the rook cluster
- Look at the osd pod logs

**Comment by BlaineEXE** [comment]
I'm unable to repro this with Rook master and Ceph 13.2.2. Is this an issue with Rook 0.9.1?

**Comment by travisn** [comment]
Do you see any output in the osd log after the `ceph-osd` command? All I see is the output collected from ceph-volume, then the command to start ceph-osd. When debugging an issue yesterday with someone else, we found that setting the `log file` in the ceph config to a file instead of `/dev/stderr` gave us a bunch of output to the file

**Comment by BlaineEXE** [comment]
Here is a pastebin of the log from one of my osd containers. https://paste.opensuse.org/view//1664407a

It seems to me that this is what I would expect, but maybe there is data missing I don't see. My pods are running bluestore with `ceph-osd` as the entrypoint. It sounds like you could be using filestore.

**Comment by noahdesu** [comment]
I think the issue here is that the osd spec should use `-d` instead of `--foreground`. verifying...

**Comment by travisn** [comment]
@noahdesu Yes, collecting all the log output would be the goal. Files inside the pod are more easily lost when the pod is restarted.

**Comment by liewegas** [comment]
> The logging problem might just be that the default logging level is 0, which seems far too low. Seems like we should remove these [default logging levels](https://github.com/rook/rook/blob/master/pkg/daemon/ceph/config/config.go#L213) and rely on the default ceph logging levels. @liewegas thoughts on this?

Yes, please just use the ceph defaults.

**Comment by travisn** [comment]
The issue seems to be in the following scenario, only affecting a certain type of osd:
- On builds where `ceph-volume` is available (12.2.10, 13.2.4, and newer), rook provisions the osd
- Rook starts the osd pod with the entrypoint of `rook`.
- Rook creates [two processes](https://github.com/rook/rook/blob/master/pkg/daemon/ceph/osd/daemon.go#L42):
  - `ceph-volume activate`, which creates a tmpfs for the config directory on the volume
  - `ceph-osd` to run the osd daemon
- Rook attempts to capture all the stdout and stderr output from the two processes and [writes them to the log](https://github.com/rook/rook/blob/master/pkg/daemon/ceph/osd/daemon.go#L42)

Rook seems to be losing the stderr that is written by the `ceph-osd` process. It is certainly related to mounting `/dev` into the container for the osds to access and the output to `/dev/stderr` must be confused. However, it's not clear where the disconnect is in the logging.

The workaround to collect osd logs is to set the `log file` config option to write to a file in the container instead of stderror. For example, set a [custom ceph.conf](https://rook.io/docs/rook/v0.9/advanced-configuration.html#custom-cephconf-settings) setting as follows:
```
[global]
log file = /var/lib/rook/$name.log
```

**Comment by zerkms** [comment]
It looks like this problem is relevant to mine https://github.com/rook/rook/issues/2547

It looks like the `/rook/rook` process after it starts `ceph-osd` it does not forward the `stderr` wherever it should and after couple minutes of running under IO load - some of its buffers fills and `ceph-osd` blocks forever.

**Comment by noahdesu** [comment]
@zerkms https://github.com/noahdesu/rook/commit/8f13ccf102854339b7214b93c79d3cd4566356d2

**Comment by zerkms** [comment]
@noahdesu exactly. It's curious the problem was not spotted for more than 2 years o_O

**Comment by zerkms** [comment]
@noahdesu right, I did not spot a PR was already sent to address it, sorry.

**Comment by jbw976** [comment]
Does this need to be backported to 0.9? it's in the 0.9 project.

**Comment by noahdesu** [comment]
yes
